File: built-in plugins dsp/mb-phaser-classic.cpp
/**
 * MB Classic Phaser
 * Category : effect
 * Type     : phaser
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : 4-stage phaser
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_PHASER_CLASSIC_H
#define MB_PHASER_CLASSIC_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbPhaserClassic : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-phaser-classic";
    static constexpr const char* PLUGIN_NAME    = "MB Classic Phaser";
    static constexpr const char* PLUGIN_TYPE    = "phaser";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float rate = 0.5f;  // range [0.01, 10]
    float depth = 0.7f;  // range [0, 1]
    float feedback = 0.5f;  // range [0, 0.99]
    float mix = 0.5f;  // range [0, 1]
    };

    MbPhaserClassic() = default;
    ~MbPhaserClassic() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.rate = std::clamp(params.rate, 0.01f, 10f);
        params.depth = std::clamp(params.depth, 0f, 1f);
        params.feedback = std::clamp(params.feedback, 0f, 0.99f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Classic Phaser
        return input;
    }
};

#endif // MB_PHASER_CLASSIC_H

File: built-in plugins dsp/mb-phaser-classic.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbPhaserClassicPlugin: PluginDefinition = { id: 'mb-phaser-classic', slug: 'mb-phaser-classic', name: 'MB Classic Phaser', category: 'effect', type: 'phaser', version: '1.0.0', description: '4-stage phaser', author: 'Max Booster', grade: 'A', parameters: [{ id: 'rate', name: 'Rate', type: 'float', defaultValue: 0.5, minValue: 0.01, maxValue: 10, automatable: true }, { id: 'depth', name: 'Depth', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true }, { id: 'feedback', name: 'Feedback', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 0.99, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { rate: 0.5, depth: 0.7, feedback: 0.5, mix: 0.5 } };

export default MbPhaserClassicPlugin;
